src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs

```diff
@@ -1264,7 +1269,8 @@ private sealed class LoaderHolder
                 // Setting the defaults to the user provided values will avoid these in the output of the call CmdParser.GetSettings.
                 UseThreads = options.UseThreads,
                 HeaderFile = options.HeaderFile,
-                MaxRows = options.MaxRows
+                MaxRows = options.MaxRows,
+                HasHeader = options.HasHeader
```

Comment:
>HasHeader [](start = 36, length = 9)

Why does this need to change?

Reply:
If `headerFile` is not provided, and we are trying to parse the schema from `dataSample` here:
https://github.com/dotnet/machinelearning/blob/f0a8a76a40348021b5a24f9b3027d53fff4a58cd/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs#L1118

And if `hasHeader` is also true, then `"header=+"` is added to `tmp` here:
https://github.com/dotnet/machinelearning/blob/f0a8a76a40348021b5a24f9b3027d53fff4a58cd/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs#L1261-L1268

Then the schema is not read from `dataSample` and this warning is thrown here:
https://github.com/dotnet/machinelearning/blob/f0a8a76a40348021b5a24f9b3027d53fff4a58cd/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs#L1280-L1285

Reply:
the call to to `CmdParser.GetSettings(...)` in line 1266, gets a string that includes all the arguments that are different between `options` (the user specified options) and
```
new Options() { ... }
```

Since what this method does is try to create an `Options` object from the schema in the file, we want to make sure that the user didn't specify any non-default arguments, because if they did - we use what the user wants instead of what the file has. However, there are a few arguments where we will still read the `Options` from the file even if they have non-default values, since they can differ each time you load the file:
* `UseThreads` - since if you load the file on a different machine you might or might not want to use threads
* `HeaderFile` - since if the header is in a separate file the user will need to specify it when the loader is created, depending on where it is located
* `MaxRows` - since at different times you might want to load just a certain number of rows 

The `HasHeader` argument is not one of the arguments that the user is allowed to change between runs, since the file either has a header or not, and if it does then the `Options` defined in the file already contains "header=+". So this code change should be reverted.


---
In reply to: [385412117](https://github.com/dotnet/machinelearning/pull/4899#discussion_r385412117) [](ancestors = 385412117)

Reply:
So does `hasHeader` refer to the `dataSample` file and not the data file being loaded? If that is the case, I will update the documentation to reflect that.

As a user, I might think that `hasHeader` refers to the data I will load with `loader.Load` and want to skip the header there. But if I set it to true, and also provide a sample, and the sample   has header in the schema but not "header=+", the schema will not be read.

Reply:
`hasHeader` is used for getting slot names from `dataSample`. It is used for determining whether to skip the first line in the file specified in `loader.Load`. 
The `hasHeader` argument does not refer to the schema that is sometimes present in the file, it only refers to a first line in the file that contains slot names. 

---
In reply to: [385861537](https://github.com/dotnet/machinelearning/pull/4899#discussion_r385861537) [](ancestors = 385861537)